# RNA sequence tutorial

### This tutorial is perfect if:

* you are just starting bioinformatics and you want to learn how to install necessary softwares
* you want to replicate someone else's data in-house
* you are currently paying to convert fastq into read count files but you want to do it for free
* you have many samples you want to analyze and you want to do it fast & automatically

We will use *Drosophila melanogaster* as the model organism. The modules are in these sequences:

#### 1. Basic installation guide
#### 2. Learn how to convert 1 fastq into 1 read count file
#### 3. Learn how to convert multiple fastq into 1 read count file
#### 4. Learn how to conduct simple downstream analysis


### Visit the [Wiki page](https://github.com/naoto-hikawa/RNAseq_tutorial/wiki) now to get started!

***

***Disclaimer***

This tutorial was motivated by my own experience of struggling to build an RNA-seq pipeline, and is not intended to be the "correct" solution for everything; I used the combination of STAR align + featureCounts for the fast processing. My preferred coding language is Java, so if you find any mistakes in Python, R, or bash scripts I provided, please let me know! 
